src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -0,0 +1,201 @@
+      SnowflakeStreamingIngestClient client) {
+    return new StreamingClientProvider(createdClientId, connectorConfig, client);
+  }
+
+  private static final KCLogger LOGGER = new KCLogger(Utils.class.getName());
+  private static final String STREAMING_CLIENT_PREFIX_NAME = "KC_CLIENT_";
+
+  private int createdClientId;
+  private Map<String, String> connectorConfig;
+  private SnowflakeStreamingIngestClient streamingIngestClient;
+
+  // private constructor for singleton
+  private StreamingClientProvider() {
+    this.createdClientId = 0;
+    this.connectorConfig = new HashMap<>();
+  }
+
+  // ONLY FOR TESTING - private constructor to inject properties for testing
+  @VisibleForTesting
+  private StreamingClientProvider(
+      int createdClientId,
+      Map<String, String> connectorConfig,
+      SnowflakeStreamingIngestClient client) {
+    this.createdClientId = createdClientId;
+    this.connectorConfig = connectorConfig;
+    this.streamingIngestClient = client;
+  }
+
+  /**
+   * Creates a new streaming client, will replace the existing one if necessary
+   *
+   * @param connectorConfig The connector config to define the client
+   */
+  public void createClient(Map<String, String> connectorConfig) {
+    // replace previous connector config and client if applicable
+    LOGGER.info("Creating new client, this will replace the old client and config if exists");
+    this.closeClient();
+
+    this.connectorConfig = connectorConfig;
+    this.streamingIngestClient = this.initStreamingClient(this.connectorConfig);
+  }
+
+  /** Closes the current client */
+  public void closeClient() {
+    // don't do anything if client is already invalid
+    if (!isClientValid(this.streamingIngestClient)) {
+      LOGGER.info("Streaming client already closed");
+      return;
+    }
+
+    LOGGER.info("Closing Streaming Client:{}", this.streamingIngestClient.getName());
+    try {
+      this.streamingIngestClient.close();
+    } catch (Exception e) {
+      // the client should auto close, so don't throw an exception here
+      String message =
+          e.getMessage() == null || e.getMessage().isEmpty()
+              ? "missing exception message"
+              : e.getMessage();
+      String cause =
+          e.getCause() == null || e.getCause().getStackTrace() == null
+              ? "missing exception cause"
+              : Arrays.toString(e.getCause().getStackTrace());
+
+      LOGGER.error("Failure closing Streaming client msg:{}, cause:{}", message, cause);
+    }
+  }
+
+  /**
+   * Gets the current client or creates a new one from the given connector config If client
+   * optimization is not enabled, just create a new streaming client
+   *
+   * @param connectorConfig The connector config, given as a backup in case the current client is
+   *     invalid
+   * @return The current or newly created streaming client
+   */
+  public SnowflakeStreamingIngestClient getClient(Map<String, String> connectorConfig) {
```

Comment:
I think this method should be threadsafe. We should also make all create and close client method thread safe.

Reply:
It didnt need to be threadsafe because every task would spin up a SnowflakeSinkServiceV2 instance which init a client before. 

If we are doing this at a singleton level, it needs to be threadsafe.

WDYT? 

Can you also add a test where in two separate threads call the `getClient` method?

Reply:
I agree it should be threadsafe, I've added a reentrant lock to ensure that it is. Will add a couple unit tests next

Reply:
Added unit tests, let me know how it looks